### ISSUE TITLE
AJAX sign up

### ISSUE THREAD

**Issue by moondawg69** [clearbold/craft-campaignmonitor-lists#7]
Would you happen to know if this sign up form would work with the Campaign Monitor old Ajax sign up form code? If one just added the ID to your form?

<script type="text/javascript">
    $(function () {
        $('#subForm').submit(function (e) {
            e.preventDefault();
            $.getJSON(
            this.action + "?callback=?",
            $(this).serialize(),
            function (data) {
                if (data.Status === 400) {
                    alert("Error: " + data.Message);
                } else { // 200
                    alert("Success: " + data.Message);
                }
            });
        });
    });
</script>

**Comment by heymarkreeves** [comment]
Hello!

It's not going to behave exactly the same, but you can submit your subscribe form via Ajax so that your design stays the same. Issue #3 addressed things working with Ajax.

Mark

**Comment by moondawg69** [comment]
Mark - many thanks for your reply and your work on this. I am just giving your sample code a run, before trying the AJAX side. 

The form is submitting, refreshes the page - no success message etc, and the new subscriber is not registering in the plugin view or in CM admin itself on the subscriber list.

Latest Craft, and code below - I am assuming the listID I have to hardcode?

`<form method="post" action="" accept-charset="UTF-8" class="padding-top-4 small-form">
      {{ csrfInput() }}
      <input type="hidden" name="action" value="cm-lists/subscribe" />
      <input type="hidden" name="redirect" value="{{ ''|hash }}" />
      <input type="hidden" name="listId" value="{{ 'e92d61cfc728255e2540ef23f30feded'|hash }}" />
<div class="input-group">
<input class="input-group-field" type="text" placeholder="email address" type="email" name="email">
<div class="input-group-button">
<input type="submit" class="group-button" value="SIGN UP">
</div>`

What am I doing wrong? TIA.

**Comment by heymarkreeves** [comment]
You'll need to first make sure you have your API key & Client ID installed in the Campaign Monitor Service plugin.

Then, per the sample code, you need to encode the list ID in the form.

Your code didn't come through the email. I'm not sure how the email came through—there's a "view on GitHub" link, but that's just the closed issue.

**Comment by moondawg69** [comment]
Hi Mark,

Once again, thank you for replying. I got everything working, thank you. Great plugin.

One thing that is doing this old man's head in, the AJAX is working, apart from actually unhiding a message. 

Form:

```
<form method="post" action="" accept-charset="UTF-8" class="padding-top-4 small-form" id="pan-form">
      {{ csrfInput() }}
      <input type="hidden" name="action" value="cm-lists/subscribe" />
      <input type="hidden" name="redirect" value="{{ ''|hash }}" />
      <input type="hidden" name="listId" value="{{ '79adce7db19b885643819f73b83e0d91'|hash }}" />
<div class="input-group" id="formfields">
<input class="input-group-field" placeholder="email address" type="email" name="email">
<div class="input-group-button">
<input type="submit" class="group-button" value="Subscribe" id="submit-button">

</div>

</div>
<p id="cheers" style="display: none;">Thank you, you are now subscribed</p>
```

And the JS:

```
$('#pan-form').submit(function(ev) {
// Prevent the form from actually submitting
ev.preventDefault();

// Send it to the server
$.post({
url: '/',
dataType: 'json',
data: $(this).serialize(),
success: function(response) {
if (response.success) {
    $('#cheers').show();
} else {
    // response.error will be an object containing any validation errors that occurred, indexed by field name
    // e.g. response.error.fromName => ['From Name is required']
    alert('An error occurred. Please try again.');
}
}
});
});
```

Does the plugin not allow me to target the response success to trigger the change?

2 hours now lol...

**Comment by heymarkreeves** [comment]
The Campaign Monitor response bundle is returned as JSON. If you console.log that, you'll see what you can parse out. Trying submitting with both valid and invalid list IDs to see what the different responses look like. But it's pretty much just passing through what you'd get from the Campaign Monitor RESTful API.

- Mark

On Sat, Jan 11, 2020, at 2:17 PM, Jez Timms wrote:
> Hi Mark,

> Once again, thank you for replying. I got everything working, thank you. Great plugin.

> One thing that is doing this old man's head in, the AJAX is working, apart from actually unhiding a message.

> Form:

> `<form method="post" action="" accept-charset="UTF-8" class="padding-top-4 small-form" id="pan-form">
      {{ csrfInput() }}
      <input type="hidden" name="action" value="cm-lists/subscribe" />
      <input type="hidden" name="redirect" value="{{ ''|hash }}" />
      <input type="hidden" name="listId" value="{{ '79adce7db19b885643819f73b83e0d91'|hash }}" />
<div class="input-group" id="formfields">
<input class="input-group-field" placeholder="email address" type="email" name="email">
<div class="input-group-button">
<input type="submit" class="group-button" value="Subscribe" id="submit-button">

</div>

</div>
<p id="cheers" style="display: none;">Thank you, you are now subscribed</p>
`
> And the JS:

> `$('#pan-form').submit(function(ev) {
// Prevent the form from actually submitting
ev.preventDefault();

// Send it to the server
$.post({
url: '/',
dataType: 'json',
data: $(this).serialize(),
success: function(response) {
if (response.success) {
    $('#cheers').show();
} else {
    // response.error will be an object containing any validation errors that occurred, indexed by field name
    // e.g. response.error.fromName => ['From Name is required']
    alert('An error occurred. Please try again.');
}
}
});
});
`

**Comment by moondawg69** [comment]
Ah thank you. I just noticed I cannot use the CM custom multi choice fields at present in the plugin which I need to segment campaigns. Think I'll have to go the Freeform route.

Thank you again for your time.

**Comment by heymarkreeves** [comment]
Want to post that as a new issue as a feature request? You might want to try setting up a series of checkbox fields named as an array. Something like name="checkboxField[]" on each. If those get parsed as one field in PHP automatically, they might get passed over correctly to Campaign Monitor. If not, if that's something available in CM, it's something I should look at adding support for.

- Mark

**Comment by moondawg69** [comment]
Very kind Mark thank you. I couldn’t even get the simple AJAX to work :(

I have done this many times with the Craft Contact Form and I am flummoxed as to why nothing happens. I assume the form posts to your plugin, the plugin gets a response from the CM API and the plugin returns a success JSON so I could then use my JS. That doesn't seem to be the case though? Sorry, but going mad ;)
